xFileSystemAccessRule: Make Cluster Aware

randomnote1 opened the pull request:
**Pull Request (PR) description**
Made the resource cluster aware. When ProcessOnlyOnActiveNode is specified, the resource will only determine if a change is needed if the target node is the active host of the file system object.

**This Pull Request (PR) fixes the following issues:**
Fixes #16

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/powershell/xsystemsecurity/18)
<!-- Reviewable:end -->

johlju commented:
@randomnote1 Can you work on this if I review?

randomnote1 commented:
@johlju Yes, I am ready to go. Thanks for reviewing!

johlju commented:
@randomnote1 Great! I come back to this tomorrow. I have the e-mail notification as reminder.

johlju commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[README.md, line 37 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCgsWWG4Ugvia12izls:-LCgsWWG4Ugvia12izlt:b-uqmz95) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/README.md#L37)):*
> ```
> can be empty if ensure = absent
> ```

Maybe `..rule. Optional if Ensure is set to the value 'Absent'. Default value is an empty string array.`

---

*[README.md, line 38 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCgsYFh-d244P9ktGiC:-LCgsYFh-d244P9ktGiD:bc2c1xy) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/README.md#L38)):*
> ```Markdown
> * **`[String]` Identity** _(Key)_: The identity to set permissions for
> * **`[String[]]` Rights** _(Write)_: The permissions to include in this rule, can be empty if ensure = absent. Should be a combination of the following:
>   * "ListDirectory"
> ```

Could we add these like this instead?

{ *ListDirectory*  | ReadData | WriteData | ...etc... }`

---

*[README.md, line 61 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCgsCMt81ZCD79T0z-3:-LCgsCMt81ZCD79T0z-4:biuobjz) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/README.md#L61)):*
> ```Markdown
>   * "Synchronize"
>   * "FullControl"
> * **`[String]` Ensure** _(Write)_: Present to create the rule, Absent to remove an existing rule
> ```

Can we add this?  `...existing rule. Default value is 'Present'. { *Present*  | Absent }`

---

*[README.md, line 62 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCgwGTQDqfkLuNFhb4X:-LCgwGTQDqfkLuNFhb4Y:b584ubx) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/README.md#L62)):*
> ```Markdown
>   * "FullControl"
> * **`[String]` Ensure** _(Write)_: Present to create the rule, Absent to remove an existing rule
> * **`[Boolean]` ProcessOnlyOnActiveNode** _(Write)_: Specifies that the resource will only determine if a change is needed if the target node is the active host of the filesystem object.
> ```

Should we add that to use this the user the configuration is run as must have permission in the cluster?

---

*[README.md, line 63 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCgzMw69xOrZ6MRDp-g:-LCgzMw69xOrZ6MRDp-h:b-r6laja) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/README.md#L63)):*
> ```Markdown
> * **`[String]` Ensure** _(Write)_: Present to create the rule, Absent to remove an existing rule
> * **`[Boolean]` ProcessOnlyOnActiveNode** _(Write)_: Specifies that the resource will only determine if a change is needed if the target node is the active host of the filesystem object.
> * **`[Boolean]` IsActiveNode** _(Read)_: Determines if the current node is actively hosting the filesystem object.
> ```

Maybe add something like this (please proof read): "This will always return $true if ProcessOnlyOnActiveNode is left out or set to the value $false."

---

*[README.md, line 69 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCgtCOx6aX5mIG1vdiZ:-LCgtCOx6aX5mIG1vdi_:b-5qqtya) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/README.md#L69)):*
> ```Markdown
> ## Versions
> 
> ### Unreleased
> ```

Please add an entry or entries for the changes to the Unreleased section.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 38 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCgukpq55kagWKU2jPt:-LCgukpq55kagWKU2jPu:b-rylbdo) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L38)):*
> ```PowerShell
> 
>         # Is the node a member of a WSFC?
>         $msCluster = Get-CimInstance -Namespace root/MSCluster -ClassName MSCluster_Cluster -ErrorAction SilentlyContinue
> ```

If the resource is run with a user that does not have permission to evaluate the cluster, this would just continue without throwing an error. Can we catch that, or will the note in the README.md be sufficient? If we should catch that, then maybe we should only do the cluster parts if `ProcessOnlyOnActiveNode -eq $true`?

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 82 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh0Xu6C6F-36lHSaYi:-LCh0Xu6C6F-36lHSaYj:b-jm9rnd) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L82)):*
> ```PowerShell
>         $accessRules = $acl.Access
> 
>         $result.Rights = @(
> ```

This will return an object array of type FileSystemRights. Not a string array?
```
PS > $rights.GetType()

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     Object[]                                 System.Array


PS > $rights[0].GetType()

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     FileSystemRights                         System.Enum
```

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 91 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh2o1FA1SiMYcRgcsk:-LCh2o1FA1SiMYcRgcsl:b5c1puh) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L91)):*
> ```PowerShell
> }
> 
> <#
> ```

Update the comment-based help with the added/changed descriptions in the README.md.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 151 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh1M908DGSBMMRPeMM:-LCh1M908DGSBMMRPeMN:bjfaw03) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L151)):*
> ```PowerShell
>         )]
>         [String[]]
>         $Rights = @(),
> ```

The default value here, does that mean that when used with 'Present' that will replace all existing permissions with no permissions for that identity, or will it not do anything?

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 200 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh2eJ35K5aZO3Tqypx:-LCh2eJ35K5aZO3Tqypy:b-mumyaa) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L200)):*
> ```
> -Verbose
> ```

We should not have `-Verbose` here?

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 319 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh3k3o23smwbaq6iMj:-LCh3k3o23smwbaq6iMk:b-ggje9x) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L319)):*
> ```PowerShell
>     if ( $comparisonResult.Count -gt 0 )
>     {
>         Write-Verbose -Message ( 'The identity "{0}" has the rights "{1}".' -f $Identity,( $currentValues.Rights -join ', ' ) ) -Verbose
> ```

Maybe write out what they should be too, the expected rights?

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.schema.mof, line 7 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCgtXYqBHqYTh9Jq5-c:-LCgtXYqBHqYTh9Jq5-d:bdfdciy) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.schema.mof#L7)):*
> ```
>     [Key, Description("The identity to set permissions for")] string Identity;
>     [Write, Description("The permissions to include in this rule, can be empty if ensure = absent"), ValueMap{"ListDirectory","ReadData","WriteData","CreateFiles","CreateDirectories","AppendData","ReadExtendedAttributes","WriteExtendedAttributes","Traverse","ExecuteFile","DeleteSubdirectoriesAndFiles","ReadAttributes","WriteAttributes","Write","Delete","ReadPermissions","Read","ReadAndExecute","Modify","ChangePermissions","TakeOwnership","Synchronize","FullControl"}, Values{"ListDirectory","ReadData","WriteData","CreateFiles","CreateDirectories","AppendData","ReadExtendedAttributes","WriteExtendedAttributes","Traverse","ExecuteFile","DeleteSubdirectoriesAndFiles","ReadAttributes","WriteAttributes","Write","Delete","ReadPermissions","Read","ReadAndExecute","Modify","ChangePermissions","TakeOwnership","Synchronize","FullControl"}] string Rights[];
>     [Write, Description("Present to create the rule, Absent to remove an existing rule"), ValueMap{"Present","Absent"}, Values{"Present","Absent"}] string Ensure;
> ```

Please match the changes in parameter descriptions to the descriptions here (those you have changed, or will change) :slightly_smiling_face: 

---

*[Tests/Unit/MSFT_xFileSystemAccessRule.tests.ps1, line 13 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh4BXFDYKUchZGA0wW:-LCh4BXFDYKUchZGA0wX:b-h2rw84) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/Tests/Unit/MSFT_xFileSystemAccessRule.tests.ps1#L13)):*
> ```PowerShell
> Import-Module -Name (Join-Path -Path $script:moduleRoot -ChildPath (Join-Path -Path 'DSCResource.Tests' -ChildPath 'TestHelper.psm1')) -Force
> 
> # TODO: Insert the correct <ModuleName> and <ResourceName> for your resource
> ```

Let's remove the TODO comment.

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xsystemsecurity/18)*
<!-- Sent from Reviewable.io -->

johlju commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 42 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh62cM8emGmz3pHYHJ:-LCh62cM8emGmz3pHYHK:b-hshsqm) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L42)):*
> ```
> -Verbose
> ```

We should not need this here.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 68 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh6-M_-qIShKAHdLlm:-LCh6-M_-qIShKAHdLln:b-hshsqm) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L68)):*
> ```
> -Verbose
> ```

We should not need this here.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 173 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh6AQnDjg9JZzacthG:-LCh6AQnDjg9JZzacthH:b-hshsqm) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L173)):*
> ```
> -Verbose
> ```

We should not need this here.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 294 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh6E6qDPKoIiOFNhpI:-LCh6E6qDPKoIiOFNhpJ:b-hshsqm) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L294)):*
> ```
> -Verbose
> ```

We should not need this here.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 319 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh6Gar3p-qBXJ-btdq:-LCh6Gar3p-qBXJ-btdr:b-hshsqm) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L319)):*
> ```
> -Verbose
> ```

We should not need this here.

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xsystemsecurity/18)*
<!-- Sent from Reviewable.io -->

randomnote1 commented:
Review status: 3 of 4 files reviewed at latest revision, 19 unresolved discussions.

---

*[README.md, line 37 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCgsWWG4Ugvia12izls:-LChnMB-7jFheqeYxEwN:b-896fix) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/README.md#L37)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> can be empty if ensure = absent
> ```

Maybe `..rule. Optional if Ensure is set to the value 'Absent'. Default value is an empty string array.`
</blockquote></details>

Done.

---

*[README.md, line 38 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCgsYFh-d244P9ktGiC:-LChoPr45PRW2pLqWE30:b-896fix) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/README.md#L38)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Could we add these like this instead?

{ *ListDirectory*  | ReadData | WriteData | ...etc... }`
</blockquote></details>

Done.

---

*[README.md, line 61 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCgsCMt81ZCD79T0z-3:-LChogrp-fViQSmIYSYR:b-896fix) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/README.md#L61)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Can we add this?  `...existing rule. Default value is 'Present'. { *Present*  | Absent }`
</blockquote></details>

Done.

---

*[README.md, line 62 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCgwGTQDqfkLuNFhb4X:-LChorYG-aSKchrvJLQN:b-896fix) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/README.md#L62)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Should we add that to use this the user the configuration is run as must have permission in the cluster?
</blockquote></details>

Done.

---

*[README.md, line 63 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCgzMw69xOrZ6MRDp-g:-LChphF84BFcuKrpMSjs:b-896fix) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/README.md#L63)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Maybe add something like this (please proof read): "This will always return $true if ProcessOnlyOnActiveNode is left out or set to the value $false."
</blockquote></details>

Done.

---

*[README.md, line 69 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCgtCOx6aX5mIG1vdiZ:-LChreiY2bhNk9xdXwPW:b-896fix) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/README.md#L69)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Please add an entry or entries for the changes to the Unreleased section.
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 38 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCgukpq55kagWKU2jPt:-LCi8rBN7PE-VfZFn29F:bl6ifxi) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L38)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

If the resource is run with a user that does not have permission to evaluate the cluster, this would just continue without throwing an error. Can we catch that, or will the note in the README.md be sufficient? If we should catch that, then maybe we should only do the cluster parts if `ProcessOnlyOnActiveNode -eq $true`?
</blockquote></details>

I can think of a couple options here:

1) Trap the permissions failure message and silently continue if the cluster doesn't exist
1) Leave it as-is because the failure will ultimately be caught in the `throw "Unable to get ACL for '$Path' because it does not exist"` line.

I'm leaning toward the second option because I don't think adding the additional error handling adds much value.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 42 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh62cM8emGmz3pHYHJ:-LChtCog04Gvyyyvz5pY:b-896fix) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L42)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> -Verbose
> ```

We should not need this here.
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 68 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh6-M_-qIShKAHdLlm:-LChtJzu4KfFbNV5P6vS:b-896fix) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L68)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> -Verbose
> ```

We should not need this here.
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 82 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh0Xu6C6F-36lHSaYi:-LChvLdBA22y_UNpvgwC:b-jkybnb) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L82)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

This will return an object array of type FileSystemRights. Not a string array?
```
PS > $rights.GetType()

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     Object[]                                 System.Array


PS > $rights[0].GetType()

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     FileSystemRights                         System.Enum
```
</blockquote></details>

No, it returns a string array. I added a type to the variable to make readability easier.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 91 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh2o1FA1SiMYcRgcsk:-LChw6iS5PdAOJdWMdwh:b-896fix) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L91)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Update the comment-based help with the added/changed descriptions in the README.md.
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 151 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh1M908DGSBMMRPeMM:-LChwX0kCR8oFFaHHD1d:b-wbk3vf) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L151)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

The default value here, does that mean that when used with 'Present' that will replace all existing permissions with no permissions for that identity, or will it not do anything?
</blockquote></details>

Yes, good point. Since `$Rights` is optional when `$Ensure = 'Absent'`, I added a test to determine if the `$Rights` parameter was passed when `$Ensure = 'Present'`.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 173 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh6AQnDjg9JZzacthG:-LChtRrY9xFQLfzAx-md:b-896fix) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L173)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> -Verbose
> ```

We should not need this here.
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 200 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh2eJ35K5aZO3Tqypx:-LChtUpm7N4jNISxUY01:b-896fix) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L200)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> -Verbose
> ```

We should not have `-Verbose` here?
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 294 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh6E6qDPKoIiOFNhpI:-LChtXXKBpdNGyhdGtqg:b-896fix) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L294)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> -Verbose
> ```

We should not need this here.
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 319 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh6Gar3p-qBXJ-btdq:-LCht_7sAORtp9Y8G8EV:b-896fix) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L319)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> -Verbose
> ```

We should not need this here.
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 319 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh3k3o23smwbaq6iMj:-LChzKCuENupgFIFmgLI:b-896fix) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L319)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Maybe write out what they should be too, the expected rights?
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.schema.mof, line 7 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCgtXYqBHqYTh9Jq5-c:-LCi-cff3o0a34KJNJPm:b-896fix) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.schema.mof#L7)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Please match the changes in parameter descriptions to the descriptions here (those you have changed, or will change) :slightly_smiling_face: 
</blockquote></details>

Done.

---

*[Tests/Unit/MSFT_xFileSystemAccessRule.tests.ps1, line 13 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh4BXFDYKUchZGA0wW:-LChtnkU6l-5SKlvQEd-:b-896fix) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/Tests/Unit/MSFT_xFileSystemAccessRule.tests.ps1#L13)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Let's remove the TODO comment.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xsystemsecurity/18)*
<!-- Sent from Reviewable.io -->

randomnote1 commented:
@johlju The first round of updates are complete.

johlju commented:
Awesome work again! Just a few tiny comments left! :slightly_smiling_face: 

---

Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[README.md, line 39 at r3](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCn-SlB4IZjfdDy2rvn:-LCn-SlB4IZjfdDy2rvo:b-a8wul7) ([raw file](https://github.com/powershell/xsystemsecurity/blob/17e4286805633313a260dafb441b4695e81172ed/README.md#L39)):*
> ```
> the user
> ```

The (upper 'T')

---

*[README.md, line 39 at r3](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCn-Xk237vU4fcpVRFH:-LCn-Xk237vU4fcpVRFI:b9ym3mb) ([raw file](https://github.com/powershell/xsystemsecurity/blob/17e4286805633313a260dafb441b4695e81172ed/README.md#L39)):*
> ```
> haver
> ```

have

---

*[README.md, line 49 at r3](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCn-qiJD2E-hg6UBNiy:-LCn-qiJD2E-hg6UBNiz:b-kedh9k) ([raw file](https://github.com/powershell/xsystemsecurity/blob/17e4286805633313a260dafb441b4695e81172ed/README.md#L49)):*
> ```
> [Dan Reist (@randomnote1)](https://github.com/randomnote1)
> ```

In other resource modules we switched to have the contributor name at the end of the change log entry instead, to make reading of the change log easier. Can we do that here to?

```
  * Fixed issue when cluster shared disk is not present on the server ([issue #16](https://github.com/PowerShell/xSystemSecurity/issues/16)). [Dan Reist (@randomnote1)](https://github.com/randomnote1)
```

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 38 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCgukpq55kagWKU2jPt:-LCn0qRPBT_FKXS3KgLI:buvgz8p) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L38)):*
<details><summary><i>Previously, randomnote1 (Dan Reist) wrote…</i></summary><blockquote>

I can think of a couple options here:

1) Trap the permissions failure message and silently continue if the cluster doesn't exist
1) Leave it as-is because the failure will ultimately be caught in the `throw "Unable to get ACL for '$Path' because it does not exist"` line.

I'm leaning toward the second option because I don't think adding the additional error handling adds much value.
</blockquote></details>

Let's go with as-is, and if there is a problem in the future that can be fixed at that time.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 82 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh0Xu6C6F-36lHSaYi:-LCn1bJ_9Vl3abFw30l3:b2gdygp) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L82)):*
> ```
> [System.String[]]@(
> ```

We should add a space between the type and @(); `[System.String[]] @()`

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 29 at r3](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCn1erbEa7QQsOOvqXH:-LCn1erbEa7QQsOOvqXI:bm36uh2) ([raw file](https://github.com/powershell/xsystemsecurity/blob/17e4286805633313a260dafb441b4695e81172ed/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L29)):*
> ```
> [System.String[]]@()
> ```

We should add a space between the type and @(); `[System.String[]] @()`

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-:-LCn2Vgw6Z9ZgNCtGbnJ:b-b0o14j)*
<!-- Sent from Reviewable.io -->

randomnote1 commented:
Review status: 1 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[README.md, line 39 at r3](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCn-SlB4IZjfdDy2rvn:-LCn3nPpE4N3c4CE1F5L:b-896fix) ([raw file](https://github.com/powershell/xsystemsecurity/blob/17e4286805633313a260dafb441b4695e81172ed/README.md#L39)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> the user
> ```

The (upper 'T')
</blockquote></details>

Done.

---

*[README.md, line 39 at r3](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCn-Xk237vU4fcpVRFH:-LCn3nssBGT6p_ca3NfE:b-896fix) ([raw file](https://github.com/powershell/xsystemsecurity/blob/17e4286805633313a260dafb441b4695e81172ed/README.md#L39)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> haver
> ```

have
</blockquote></details>

Done.

---

*[README.md, line 49 at r3](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCn-qiJD2E-hg6UBNiy:-LCn41Q9Dznl9rh52PF4:b-896fix) ([raw file](https://github.com/powershell/xsystemsecurity/blob/17e4286805633313a260dafb441b4695e81172ed/README.md#L49)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> [Dan Reist (@randomnote1)](https://github.com/randomnote1)
> ```

In other resource modules we switched to have the contributor name at the end of the change log entry instead, to make reading of the change log easier. Can we do that here to?

```
  * Fixed issue when cluster shared disk is not present on the server ([issue #16](https://github.com/PowerShell/xSystemSecurity/issues/16)). [Dan Reist (@randomnote1)](https://github.com/randomnote1)
```
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 82 at r1](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCh0Xu6C6F-36lHSaYi:-LCn4oqo14FYTs6ut-zf:b-896fix) ([raw file](https://github.com/powershell/xsystemsecurity/blob/c83a02ad4d55189b2353786ad550163b008723d7/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L82)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> [System.String[]]@(
> ```

We should add a space between the type and @(); `[System.String[]] @()`
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1, line 29 at r3](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-LCn1erbEa7QQsOOvqXH:-LCn4klc0U6NYZh_s76q:b-896fix) ([raw file](https://github.com/powershell/xsystemsecurity/blob/17e4286805633313a260dafb441b4695e81172ed/DSCResources/MSFT_xFileSystemAccessRule/MSFT_xFileSystemAccessRule.psm1#L29)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> [System.String[]]@()
> ```

We should add a space between the type and @(); `[System.String[]] @()`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xsystemsecurity/18)*
<!-- Sent from Reviewable.io -->

johlju commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xsystemsecurity/18#-:-LCnXah29VL4OVfXDjbG:bnfp4nl)*
<!-- Sent from Reviewable.io -->

johlju commented:
@randomnote1 Awesome work here! Thanks a lot for this!
